### PR TITLE
refactor: mixture density computation in multiphase poromechanics solvers

### PR DESCRIPTION
**Pull request by castelletto1**
This PR modifies the computation of the mixture (solid + fluid) density favoring use of primary variables over secondary variables.

### CODE REVIEW

**Comment by castelletto1** [comment]
We should decide whether we want to refer to the expression:

$$\rho=(1-\phi)\rho_{\text{solid}}+\phi\sum_{c}\rho_c=(1-\phi)\rho_{\text{solid}}+\phi\sum_{c}\sum_{\ell=1}^{np}x_{c\ell}\rho_{\ell}s_{\ell}=(1-\phi)\rho_{\text{solid}}+\phi\sum_{\ell=1}^{np}\rho_{\ell}s_{\ell}$$

as mixture density or bulk density, and be consistent everywhere.

**Review by paveltomin — APPROVED** [review]
❤️

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp [hunk 1]
```diff
@@ -201,27 +199,11 @@ computeBodyForce( localIndex const k,
...
-    arraySlice1d< real64 const, compflow::USD_PHASE - 1 > const phaseVolFrac = m_fluidPhaseVolFrac[k];
-    arraySlice2d< real64 const, compflow::USD_PHASE_DC - 1 > const dPhaseVolFrac = m_dFluidPhaseVolFrac[k];
-
-    // Step 1: compute fluid total mass density and its derivatives
-
-    real64 dTotalMassDensity_dTemperature = 0.0;
-
-    for( integer ip = 0; ip < m_numPhases; ++ip )
-    {
-      dTotalMassDensity_dTemperature += dPhaseVolFrac( ip, Deriv::dT ) * phaseMassDensity( ip )
```
Comment (dkachuma) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp:215]: I suppose this means that these terms cancel out and we should end up with a value that is numerically close to zero?

**Review by dkachuma — APPROVED** [review]

**Comment by CusiniM** [comment]
> We should decide whether we want to refer to the expression:
> 
> ρ = ( 1 − ϕ ) ρ solid + ϕ ∑ c ρ c = ( 1 − ϕ ) ρ solid + ϕ ∑ c ∑ ℓ = 1 n p x c ℓ ρ ℓ s ℓ = ( 1 − ϕ ) ρ solid + ϕ ∑ ℓ = 1 n p ρ ℓ s ℓ
> 
> as mixture density or bulk density, and be consistent everywhere.

can we call this `totalDensity` instead of `bulk density` ? ? It's certainly not the mixture density.

**Comment by CusiniM** [comment]
> > We should decide whether we want to refer to the expression:
> > ρ = ( 1 − ϕ ) ρ solid + ϕ ∑ c ρ c = ( 1 − ϕ ) ρ solid + ϕ ∑ c ∑ ℓ = 1 n p x c ℓ ρ ℓ s ℓ = ( 1 − ϕ ) ρ solid + ϕ ∑ ℓ = 1 n p ρ ℓ s ℓ
> > as mixture density or bulk density, and be consistent everywhere.
> 
> can we call this `totalDensity` instead of `bulk density` ? ? It's certainly not the mixture density.

Looking at the code, IMO, what we call `totalMassDensity` should be called `mixtureDensity` and the `totalDensity` should include the solid part.

**Comment by castelletto1** [comment]
> > > We should decide whether we want to refer to the expression:
> > > ρ = ( 1 − ϕ ) ρ solid + ϕ ∑ c ρ c = ( 1 − ϕ ) ρ solid + ϕ ∑ c ∑ ℓ = 1 n p x c ℓ ρ ℓ s ℓ = ( 1 − ϕ ) ρ solid + ϕ ∑ ℓ = 1 n p ρ ℓ s ℓ
> > > as mixture density or bulk density, and be consistent everywhere.
> > 
> > can we call this `totalDensity` instead of `bulk density` ? ? It's certainly not the mixture density.
> 
> Looking at the code, IMO, what we call `totalMassDensity` should be called `mixtureDensity` and the `totalDensity` should include the solid part. 

I agree it is confusing. Often, in the poromechanics literature authors refer to mixture including the solid phase, for example: _"... a three-phase mixture composed of a solid matrix whose voids are continuous and filled
with water and air ..."_ [[Borja, CMAME (2004)](https://www.sciencedirect.com/science/article/pii/S0045782504002762)].

In my opinion, we should agree on consistent naming for:
- solid phase(s)
- fluid phases and their mixture
- solid/fluid combined

Also the notation used in the documentation should be updated accordingly.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp (hunk 1 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp:215]: If we highlight in the body force term the dependencies on the primary variables we have:

$$\rho \mathbf{g}=(1-\phi(\epsilon_v(\mathbf{u}),p,T))\rho_{\text{solid}}(p,T)\mathbf{g}+\phi(\epsilon_v(\mathbf{u}),p,T)\sum_{c}\rho_c\mathbf{g}$$

We have a direct dependency on temperature exclusively on porosity at the moment -- eventually on the solid phase density $\rho_s$ (not considered at the moment in the code). I agree that one can probably show that looping over phases produces  `dTotalMassDensity_dTemperature = 0`.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp [hunk 2]
```diff
@@ -214,48 +217,26 @@ computeBodyForce( localIndex const k,
...
-                    dGlobalCompFrac_dGlobalCompDensity,
-                    dPhaseMassDensity[ip],
-                    dPhaseMassDensity_dComponents,
-                    Deriv::dC );
-    for( integer jc = 0; jc < m_numComponents; ++jc )
-    {
-      dTotalMassDensity_dComponents[jc] += dPhaseVolFrac( ip, Deriv::dC+jc ) * phaseMassDensity( ip )
-                                           + phaseVolFrac( ip ) * dPhaseMassDensity_dComponents[jc];
-    }
+    totalMassDensity += m_useMass ? compDens[ic] : compDens[ic] * m_componentMolarWeight[ic];
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp]: we should avoid `+=` in kernels.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp [hunk 3]
```diff
@@ -413,9 +416,9 @@ class MultiphaseBulkDensityKernel
                 localIndex const q ) const
   {
     m_bulkDensity[ei][q] = 0.0;
-    for( integer ip = 0; ip < m_numPhases; ++ip )
+    for( integer ic = 0; ic < m_numComponents; ++ic )
     {
-      m_bulkDensity[ei][q] += m_fluidPhaseVolFrac[ei][ip] * m_fluidPhaseDensity[ei][q][ip];
+      m_bulkDensity[ei][q] += m_useMass ? m_compDens[ei][ic] : m_compDens[ei][ic] * m_componentMolarWeight[ic];
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp]: same here. should avoid using `+=`

**Review by CusiniM — APPROVED** [review]

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp (hunk 2 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp]: Based on the offline discussion we had and on the test on Godbolt (thank you @klevzoff) this is not needed.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp (hunk 3 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp]: Same

**Comment by CusiniM** [comment]
> > > > We should decide whether we want to refer to the expression:
> > > > ρ = ( 1 − ϕ ) ρ solid + ϕ ∑ c ρ c = ( 1 − ϕ ) ρ solid + ϕ ∑ c ∑ ℓ = 1 n p x c ℓ ρ ℓ s ℓ = ( 1 − ϕ ) ρ solid + ϕ ∑ ℓ = 1 n p ρ ℓ s ℓ
> > > > as mixture density or bulk density, and be consistent everywhere.
> > > 
> > > 
> > > can we call this `totalDensity` instead of `bulk density` ? ? It's certainly not the mixture density.
> > 
> > 
> > Looking at the code, IMO, what we call `totalMassDensity` should be called `mixtureDensity` and the `totalDensity` should include the solid part.
> 
> I agree it is confusing. Often, in the poromechanics literature authors refer to mixture including the solid phase, for example: _"... a three-phase mixture composed of a solid matrix whose voids are continuous and filled with water and air ..."_ [[Borja, CMAME (2004)](https://www.sciencedirect.com/science/article/pii/S0045782504002762)].
> 
> In my opinion, we should agree on consistent naming for:
> 
>     * solid phase(s)
> 
>     * fluid phases and their mixture
> 
>     * solid/fluid combined
> 
> 
> Also the notation used in the documentation should be updated accordingly.

I mean, I think the definition of mixture is quite broad so technically one could see the solid + fluids as a mixture. However, since we are often interested in the properties of just the fluid mixture I think that it is more straightforward to call:
- `mixture` or `fluidMixture` the mixture of the fluid phases and so refer to the `mixtureDensity` or the `fluidMixtureDensity` for that one. 
- `totalDensity`  the density of the `fluidMixture` plus the `solid`. And I d do the same thing with energy, enthalpy etc. 
Mainly coz it's weird to have something called `totalDensity` that does not include all phases present.

**Comment by paveltomin** [comment]
> * `totalDensity`  the density of the `fluidMixture` plus the `solid`. And I d do the same thing with energy, enthalpy etc.
>   Mainly coz it's weird to have something called `totalDensity` that does not include all phases present.

i would say `totalDensity` is kind of reserved for fluid total density

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp (hunk 2 above)
Comment (rrsettgast) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp]: The kernel optimization efforts for matrix-free on mechanics showed that using the `+=` did not result in an FMA across the `+=` operation. This was with `cuda10.1.243`. This was confirmed by the Nvidia staff at the hackathon who conferred with some others at nvida. Maybe Nvidia fixed it...maybe it was specific to the architecture. I cannot tell you why it works for the Godbolt example, and it didn't work back then. You can disregard my guidance on this matter if you feel the lack of `+=` warrants this much effort to prove me wrong....or you could realize that I am correct:

https://godbolt.org/z/9oT8cM8T6

**Comment by CusiniM** [comment]
> > * `totalDensity`  the density of the `fluidMixture` plus the `solid`. And I d do the same thing with energy, enthalpy etc.
> >   Mainly coz it's weird to have something called `totalDensity` that does not include all phases present.
> 
> i would say `totalDensity` is kind of reserved for fluid total density

yeah, I actually don't like the name `totalDensity` for the fluid mixture one. But at the end of the day as long as we are consistent, it's fine.

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp (hunk 2 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp]: Done

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp (hunk 3 above)
Comment (castelletto1) [reply src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics.hpp]: Done

**Pull request merged**